include/vara/Feature/Error.h

```diff
@@ -0,0 +1,79 @@
+#ifndef VARA_FEATURE_ERROR_H
+#define VARA_FEATURE_ERROR_H
+
+#include "llvm/Support/raw_ostream.h"
+
+#include <cassert>
+#include <iostream>
+#include <type_traits>
+#include <variant>
+
+namespace vara::feature {
+
+enum ErrorCode {
+  ERROR,
+  ABORTED,
+  MISSING_FEATURE,
+  MISSING_PARENT,
+  ALREADY_PRESENT,
+  MISSING_MODEL,
+  INCONSISTENT
+};
+
+template <typename ValueTy, typename ErrorTy = ErrorCode,
+          std::enable_if_t<!std::is_same_v<ValueTy, ErrorTy>, bool> = true>
+class ErrorOr {
+public:
+  ErrorOr(ValueTy Value) : Value(std::move(Value)) {}
+
+  ErrorOr(ErrorTy Error) : Value(std::move(Error)) {}
+
+  /// Return false if there is an error.
+  operator bool() const { return !std::holds_alternative<ErrorTy>(Value); }
+
+  [[nodiscard]] ValueTy operator*() {
+    assert(std::holds_alternative<ValueTy>(Value));
+    return std::move(std::get<ValueTy>(Value));
+  }
+
+  [[nodiscard]] ErrorTy getError() {
+    assert(std::holds_alternative<ErrorTy>(Value));
+    return std::move(std::get<ErrorTy>(Value));
+  }
+
+protected:
+  std::variant<ValueTy, ErrorTy> Value;
+};
+
+class Result : public ErrorOr<std::monostate, ErrorCode> {
```

Comment:
Hmm, this design looks a bit unnecessary complex. Why do we need a `Result` (+not a good name) type, when we just want to return an error code? An alternative could be to define the first state of the error code to `OK`, rename it to `****State` instead (or another name) of error and just always return that type.

Maybe, we could flip this on it's head and always return a `Foo<StatusCode, Value>` and have a special `Foo<StatusCode, void>` that represents a non return function. `Foo` then specifies that Value is only correct if `StatusCode` is `OK` (similar to an optional where we check if something is in there before we are allowed to access it).